src/drivers/grpc/middlewares/auth.rs

```diff
@@ -0,0 +1,54 @@
+use std::sync::Arc;
+
+use regex::Regex;
+
+use crate::domain::users::{AuthProvider, Credential};
+
+#[derive(Clone)]
+pub struct AuthenticatorImpl {
+    auth_provider: Arc<dyn AuthProvider>,
+    token_regex: Regex,
+}
+impl AuthenticatorImpl {
+    pub fn new(auth_provider: Arc<dyn AuthProvider>) -> Self {
+        let token_regex = Regex::new(r"^(?:Bearer\s+)?(.+)$").unwrap();
+
+        Self {
+            auth_provider,
+            token_regex,
+        }
+    }
+}
+
+impl tonic::service::Interceptor for AuthenticatorImpl {
+    fn call(
+        &mut self,
+        mut request: tonic::Request<()>,
+    ) -> Result<tonic::Request<()>, tonic::Status> {
+        let Some(token) = request.metadata().get("Authorization") else {
+            return Err(tonic::Status::unauthenticated("authentication is required"));
+        };
+
+        let Ok(token) = token.to_str() else {
+            return Err(tonic::Status::unauthenticated("authentication malformed"));
+        };
+
+        let token = self
+            .token_regex
+            .captures(token)
+            .unwrap()
+            .get(1)
+            .unwrap()
+            .as_str();
```

Comment:
This will panic if the authorization header is malformed, maybe we can change it to something like:
```suggestion
        let token = match self.token_regex.captures(token) {
            Some(captures) => captures.get(1) {
                Some(token) => token.as_str(),
                None => return Err(tonic::Status::unauthenticated("Malformed authentication header"))
            },
            None => return Err(tonic::Status::unauthenticated("Malformed authentication header"))   
        }
```

Reply:
@gonzalezzfelipe Thanks for your comment, I appreciate it, but I will apply this change in the next pr because I need to change something else.